detox/src/devices/drivers/android/genycloud/GenyCloudDriver.js

```diff
@@ -0,0 +1,66 @@
+const _ = require('lodash');
+const AndroidDriver = require('../AndroidDriver');
+const GenyCloudExec = require('./exec/GenyCloudExec');
+const RecipesService = require('./services/GenyRecipesService');
+const InstanceLookupService = require('./services/GenyInstanceLookupService');
+const InstanceLifecycleService = require('./services/GenyInstanceLifecycleService');
+const InstanceNaming = require('./services/GenyInstanceNaming');
+const AllocationHelper = require('./helpers/GenyAllocationHelper');
+const DeviceQueryHelper = require('./helpers/GenyDeviceQueryHelper');
+const DetoxRuntimeError = require('../../../../errors/DetoxRuntimeError');
+const logger = require('../../../../utils/logger').child({ __filename });
+
+class GenyCloudDriver extends AndroidDriver {
+  constructor(config) {
+    super(config);
+    this._name = 'Unspecified Genymotion Cloud Emulator';
+
+    const exec = new GenyCloudExec();
+    const instanceNaming = new InstanceNaming(); // TODO should consider a permissive impl for debug/dev mode. Maybe even a custom arg in package.json (Detox > ... > genycloud > sharedAccount: false)
+    this.recipeService = new RecipesService(exec, logger);
+    this.instanceLookupService = new InstanceLookupService(exec, instanceNaming, this.deviceRegistry);
+    this.instanceLifecycleService = new InstanceLifecycleService(exec, instanceNaming);
+    this.deviceQueryHelper = new DeviceQueryHelper(this.recipeService);
+    this.allocationHelper = new AllocationHelper(this.instanceLookupService, this.instanceLifecycleService);
+  }
+
+  get name() {
+    return this._name;
+  }
+
+  async acquireFreeDevice(deviceQuery) {
+    const recipe = await this.deviceQueryHelper.getRecipeFromQuery(deviceQuery);
+    this.assertRecipe(deviceQuery, recipe);
+
+    const cookie = { coldBoot: false };
```

Comment:
Looks like we've got a serious abstraction leak in `boot` and `allocateDevice` stages, that you were forced to use a cookie object for mutations. In the previous implementations of Android drivers allocating meant only the answer for the question "shall we use a new instance, or we can reuse something?", and the booting stage was either taking control over an available device, or booting the device indeed.

Here, this increased responsibility of `allocateDevice` method gives second thoughts. I don't like this place.

Reply:
Admittedly, you are right of course. I will revise this.

Reply:
So after revisiting this today -- This is not quite true, nonetheless. The reason for the anomaly named `cookie` in this code is a result of the inflexibility of the device-registry API:

In `allocateDevice()`, the device-registry forces the user's allocation function to strictly return the device ID by which the device should be registered. This approach makes a lot of sense, but only from a naive perspective: _Allocating_ a device _does_ result in an ID, ultimately; Although the ID can also be a partial result of a more complex procedure which has side-effects. I'm not just referring to the _booting_ process, so-called, which I've arguably implemented in the _Genymotion_ allocator (and perhaps shouldn't have), but also to the very decision of whether a running device can be reused or not.

Yet while allocating a device yields such and such information, the naive API forces the function into reducing it to the final _id_ which device-registry should use, evidently resulting in information loss. The `cookie` here is therefore a means by which a function wrapping around `deviceRegistry.allocateDevice()` could both return the reduced result to the registry, and still retain the full information.

In any case, I've updated the code such that this would be simplified in upcoming PR's. However, I do think we ought to consider extending the device-registry API with something in the likes of the following suggestion:

```js
const result = deviceRegistry.performTransaction(async (addDevice, removeDevice) => {
   const deviceId = await this.somehowGetDeviceId();
   const sideEffects = {
       isNew: true,
       ...
   };
   await addDevice(deviceId);

   // Return more than just the ID
   return {
      deviceId,
      sideEffects,
   };
}

// Now result has all the information it would ever need
```
wdyt?

Reply:
@noomorph ping. I really want to be able to break the allocators down to pairs of (roughly) `reuseable_lookup`+`fresh_allocator`. I think is it a good direction. I'm currently not certain regarding whether a return value is needed at all. However I need your feedback on the approach.